libraries/AP_Math/SCurve.cpp

```diff
@@ -0,0 +1,963 @@
+        Jm_out = 0.0f;
+        t2_out = 0.0f;
+        t4_out = 0.0f;
+        t6_out = 0.0f;
+    }
+}
+
+// generate three consecutive segments forming a jerk profile
+// the index variable is the position within the path array that this jerk profile should be added
+// the index is incremented to reference the next segment in the array after the jerk profile
+void SCurve::add_segments_jerk(uint8_t &index, float tj, float Jm, float Tcj)
+{
+    add_segment_incr_jerk(index, tj, Jm);
+    add_segment_const_jerk(index, Tcj, Jm);
+    add_segment_decr_jerk(index, tj, Jm);
+}
+
+// generate constant jerk time segment
+// calculate the information needed to populate the constant jerk segment from the segment duration tj and jerk J0
+// the index variable is the position of this segment in the path array and is incremented to reference the next segment in the array
+void SCurve::add_segment_const_jerk(uint8_t &index, float tj, float J0)
+{
+    const float J = J0;
+    const float T = segment[index - 1].end_time + tj;
+    const float A = segment[index - 1].end_accel + J0 * tj;
+    const float V = segment[index - 1].end_vel + segment[index - 1].end_accel * tj + 0.5f * J0 * sq(tj);
+    const float P = segment[index - 1].end_pos + segment[index - 1].end_vel * tj + 0.5f * segment[index - 1].end_accel * sq(tj) + (1.0f / 6.0f) * J0 * powf(tj, 3.0f);
+    add_segment(index, T, SegmentType::CONSTANT_JERK, J, A, V, P);
+}
+
+// generate increasing jerk magnitude time segment based on a raised cosine profile
+// calculate the information needed to populate the increasing jerk magnitude segment from the segment duration tj and jerk magnitude Jm
+// the index variable is the position of this segment in the path array and is incremented to reference the next segment in the array
+void SCurve::add_segment_incr_jerk(uint8_t &index, float tj, float Jm)
+{
+    const float Beta = M_PI / tj;
+    const float Alpha = Jm / 2.0f;
+    const float AT = Alpha * tj;
+    const float VT = Alpha * (sq(tj) / 2.0f - 2.0f / sq(Beta));
+    const float PT = Alpha * ((-1.0f / sq(Beta)) * tj + (1.0f / 6.0f) * powf(tj, 3.0f));
+
+    const float J = Jm;
+    const float T = segment[index - 1].end_time + tj;
+    const float A = segment[index - 1].end_accel + AT;
+    const float V = segment[index - 1].end_vel + segment[index - 1].end_accel * tj + VT;
+    const float P = segment[index - 1].end_pos + segment[index - 1].end_vel * tj + 0.5f * segment[index - 1].end_accel * sq(tj) + PT;
+    add_segment(index, T, SegmentType::POSITIVE_JERK, J, A, V, P);
+}
+
+// generate decreasing jerk magnitude time segment based on a raised cosine profile
+// calculate the information needed to populate the decreasing jerk magnitude segment from the segment duration tj and jerk magnitude Jm
+// the index variable is the position of this segment in the path and is incremented to reference the next segment in the array
+void SCurve::add_segment_decr_jerk(uint8_t &index, float tj, float Jm)
```

Comment:
Does this differ from add_segment_incr_jerk *solely* by the segment type?  Maybe a minus sign buried in here?  This is a lot of duplicate code which should be factored.

Reply:
yes, perhaps it could be optimised for flash a bit but I think I'll leave it as-is to avoid the possibility of more bugs being introduced.